File: occa/device.py
import json
import numpy as np

from . import c, utils
from . import dtype as occa_dtype
from .exceptions import UninitializedError


class Device:
    def __init__(self, props=None, **kwargs):
        if isinstance(props, c.Device):
            self._c = props
            return

        utils.assert_properties(props, **kwargs)
        props = utils.properties(props, **kwargs)

        if props is not None:
            self._c = c.Device(props=props)
        else:
            self._c = None

    def _assert_initialized(self):
        if not self.is_initialized:
            raise UninitializedError('occa.Device is not initialized')

    @property
    def is_initialized(self):
        '''Return if the device has been initialized'''
        return (self._c is not None and
                self._c.is_initialized())

    def free(self):
        self._assert_initialized()
        self._c.free()

    @property
    def mode(self):
        self._assert_initialized()
        return self._c.mode()

    @property
    def properties(self):
        self._assert_initialized()
        return json.loads(self._c.properties())

    @property
    def kernel_properties(self):
        self._assert_initialized()
        return json.loads(self._c.kernel_properties())

    @property
    def memory_properties(self):
        self._assert_initialized()
        return json.loads(self._c.memory_properties())

    @property
    def memory_size(self):
        self._assert_initialized()
        return self._c.memory_size()

    @property
    def memory_allocated(self):
        self._assert_initialized()
        return self._c.memory_allocated()

    def finish(self):
        self._assert_initialized()
        self._c.finish()

    @property
    def has_separate_memory_space(self):
        self._assert_initialized()
        return self._c.has_separate_memory_space()

    #---[ Stream ]----------------------
    def create_stream(self):
        from .stream import Stream

        self._assert_initialized()
        return Stream(self._c.create_stream())

    @property
    def stream(self):
        from .stream import Stream

        self._assert_initialized()
        return Stream(self._c.get_stream())

    def set_stream(self, stream):
        self._assert_initialized()
        utils.assert_stream(stream)

        self._c.set_stream(stream._c)

    def tag_stream(self):
        from .streamtag import StreamTag

        self._assert_initialized()

        return StreamTag(self._c.tag_stream())

    def wait_for(self, tag):
        self._assert_initialized()
        utils.assert_streamtag(tag)

        self._c.wait_for(tag._c)

    def time_between(self, start, end):
        self._assert_initialized()
        utils.assert_streamtag(start)
        utils.assert_streamtag(end)

        return self._c.time_between(start._c, end._c)
    #===================================

    #---[ Kernel ]----------------------
    def build_kernel(self, filename, kernel, props=None):
        from .kernel import Kernel

        self._assert_initialized()
        utils.assert_str(filename)
        utils.assert_str(kernel)
        props = utils.properties(props) or ''

        return Kernel(
            self._c.build_kernel(filename=filename,
                                 kernel=kernel,
                                 props=props)
        )

    def build_kernel_from_string(self, source, kernel, props=None):
        from .kernel import Kernel

        self._assert_initialized()
        utils.assert_str(source)
        utils.assert_str(kernel)
        props = utils.properties(props) or ''

        return Kernel(
            self._c.build_kernel_from_string(source=source,
                                             kernel=kernel,
                                             props=props)
        )
    #===================================

    #---[ Memory ]----------------------
    def malloc(self, src_or_entries, dtype=None, props=None):
        from .memory import Memory

        self._assert_initialized()
        utils.assert_malloc_src(src_or_entries)
        props = utils.properties(props)

        src = None

        # Convert iterable into np.array
        is_ndarray = isinstance(src_or_entries, np.ndarray)
        if (not is_ndarray and
            isinstance(src_or_entries, list)):
            src_or_entries = np.array(src_or_entries)
            is_ndarray = True

        if is_ndarray:
            src = src_or_entries
            dtype = dtype or src.dtype

            entries = src.size
            if dtype != src.dtype:
                src = src.astype(dtype)
        else:
            entries = src_or_entries

        if dtype:
            dtype = np.dtype(dtype)
        utils.assert_valid_dtype(dtype)

        return Memory(
            self._c.malloc(entries=entries,
                           src=src,
                           dtype=occa_dtype.get_occa_dtype(dtype),
                           props=props),
            dtype=dtype,
        )
    #===================================

    def __bool__(self):
        return self.is_initialized

    def __eq__(self, other):
        self._assert_initialized()
        if not isinstance(other, Device):
            return False
        return hash(self) == hash(other)

    def __hash__(self):
        self._assert_initialized()
        return self._c.ptr_as_long()

File: occa/c/device.cpp
/* The MIT License (MIT)
 *
 * Copyright (c) 2018 David Medina
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 */
#include "header.hpp"


//---[ Class Python Methods ]-----------
static int Device_init(occa::py::Device *self,
                       PyObject *args,
                       PyObject *kwargs) {
  self->device = NULL;

  occa::device device;
  occa::properties props;

  occa::py::kwargParser parser;
  parser
    .startOptionalKwargs()
    .add("device", device)
    .add("props", props);

  if (!parser.parse(args, kwargs)) {
    return -1;
  }

  if (device.isInitialized()) {
    self->device = new occa::device(device);
  } else if (props.isInitialized()) {
    OCCA_INIT_TRY(
      self->device = new occa::device(props);
    );
  }

  return 0;
}

static void Device_dealloc(occa::py::Device *self) {
  delete self->device;
  Py_TYPE(self)->tp_free((PyObject*) self);
}
//======================================


//---[ Class Methods ]------------------
static PyObject* Device_is_initialized(occa::py::Device *self) {
  return occa::py::toPy(
    (bool) (self->device &&
            self->device->isInitialized())
  );
}

static PyObject* Device_free(occa::py::Device *self) {
  if (self->device) {
    self->device->free();
  }
  return occa::py::None();
}

static PyObject* Device_mode(occa::py::Device *self) {
  if (!self->device) {
    return occa::py::None();
  }
  return occa::py::toPy(
    self->device->mode()
  );
}

static PyObject* Device_properties(occa::py::Device *self) {
  if (!self->device) {
    return occa::py::None();
  }
  return occa::py::toPy(
    self->device->properties()
  );
}

static PyObject* Device_kernel_properties(occa::py::Device *self) {
  if (!self->device) {
    return occa::py::None();
  }
  return occa::py::toPy(
    self->device->kernelProperties()
  );
}

static PyObject* Device_memory_properties(occa::py::Device *self) {
  if (!self->device) {
    return occa::py::None();
  }
  return occa::py::toPy(
    self->device->memoryProperties()
  );
}

static PyObject* Device_memory_size(occa::py::Device *self) {
  if (!self->device) {
    return occa::py::None();
  }
  return occa::py::toPy(
    (long long) self->device->memorySize()
  );
}

static PyObject* Device_memory_allocated(occa::py::Device *self) {
  if (!self->device) {
    return occa::py::None();
  }
  return occa::py::toPy(
    (long long) self->device->memoryAllocated()
  );
}

static PyObject* Device_finish(occa::py::Device *self) {
  if (self->device) {
    self->device->finish();
  }
  return occa::py::None();
}

static PyObject* Device_has_separate_memory_space(occa::py::Device *self) {
  if (!self->device) {
    return occa::py::None();
  }
  return occa::py::toPy(
    self->device->hasSeparateMemorySpace()
  );
}


//  |---[ Stream ]----------------------
static PyObject* Device_create_stream(occa::py::Device *self) {
  if (!self->device) {
    return occa::py::None();
  }
  return occa::py::toPy(
    self->device->createStream()
  );
}

static PyObject* Device_get_stream(occa::py::Device *self) {
  if (!self->device) {
    return occa::py::None();
  }
  return occa::py::toPy(
    self->device->getStream()
  );
}

static PyObject* Device_set_stream(occa::py::Device *self,
                                   PyObject *args,
                                   PyObject *kwargs) {
  if (!self->device) {
    return occa::py::None();
  }

  occa::stream stream;

  occa::py::kwargParser parser;
  parser
    .add("stream", stream);

  if (!parser.parse(args, kwargs)) {
    return NULL;
  }

  self->device->setStream(stream);

  return occa::py::None();
}

static PyObject* Device_tag_stream(occa::py::Device *self) {
  if (!self->device) {
    return occa::py::None();
  }
  return occa::py::toPy(
    self->device->tagStream()
  );
}

static PyObject* Device_wait_for(occa::py::Device *self,
                                 PyObject *args,
                                 PyObject *kwargs) {
  if (!self->device) {
    return occa::py::None();
  }

  occa::streamTag tag;

  occa::py::kwargParser parser;
  parser
    .add("tag", tag);

  if (!parser.parse(args, kwargs)) {
    return NULL;
  }

  self->device->waitFor(tag);
  return occa::py::None();
}

static PyObject* Device_time_between(occa::py::Device *self,
                                     PyObject *args,
                                     PyObject *kwargs) {
  if (!self->device) {
    return occa::py::None();
  }

  occa::streamTag start, end;

  occa::py::kwargParser parser;
  parser
    .add("start", start)
    .add("end", end);

  if (!parser.parse(args, kwargs)) {
    return NULL;
  }

  return occa::py::toPy(
    self->device->timeBetween(start, end)
  );
}
//  |===================================


//  |---[ Kernel ]----------------------
static PyObject* Device_build_kernel(occa::py::Device *self,
                                     PyObject *args,
                                     PyObject *kwargs) {
  if (!self->device) {
    return occa::py::None();
  }

  std::string filename, kernel;
  occa::properties props;

  occa::py::kwargParser parser;
  parser
    .add("filename", filename)
    .add("kernel", kernel)
    .add("props", props);

  if (!parser.parse(args, kwargs)) {
    return NULL;
  }

  return occa::py::toPy(
    self->device->buildKernel(filename, kernel, props)
  );
}

static PyObject* Device_build_kernel_from_string(occa::py::Device *self,
                                                 PyObject *args,
                                                 PyObject *kwargs) {
  if (!self->device) {
    return occa::py::None();
  }

  std::string source, kernel;
  occa::properties props;

  occa::py::kwargParser parser;
  parser
    .add("source", source)
    .add("kernel", kernel)
    .add("props", props);

  if (!parser.parse(args, kwargs)) {
    return NULL;
  }

  return occa::py::toPy(
    self->device->buildKernelFromString(source, kernel, props)
  );
}

static PyObject* Device_build_kernel_from_binary(occa::py::Device *self,
                                                 PyObject *args,
                                                 PyObject *kwargs) {
  if (!self->device) {
    return occa::py::None();
  }

  std::string filename, kernel;
  occa::properties props;

  occa::py::kwargParser parser;
  parser
    .add("filename", filename)
    .add("kernel", kernel)
    .add("props", props);

  if (!parser.parse(args, kwargs)) {
    return NULL;
  }

  return occa::py::toPy(
    self->device->buildKernelFromBinary(filename, kernel, props)
  );
}
//  |===================================


//  |---[ Memory ]----------------------
static PyObject* Device_malloc(occa::py::Device *self,
                               PyObject *args,
                               PyObject *kwargs) {
  if (!self->device) {
    return occa::py::None();
  }

  long long bytes = 0;
  occa::py::ndArray src;
  occa::properties props;

  occa::py::kwargParser parser;
  parser
    .startOptionalKwargs()
    .add("bytes", bytes)
    .add("src", src)
    .add("props", props);

  if (!parser.parse(args, kwargs)) {
    return NULL;
  }

  if (!bytes) {
    bytes = src.size();
  }

  return occa::py::toPy(
    self->device->malloc(bytes,
                         src.ptr(),
                         props)
  );
}
//  |===================================
//======================================


//---[ Module ]-------------------------
#define DEVICE_METHOD_NO_ARGS(FUNC)             \
  OCCA_PY_METHOD_NO_ARGS(#FUNC, Device_##FUNC)

#define DEVICE_METHOD_WITH_KWARGS(FUNC)             \
  OCCA_PY_METHOD_WITH_KWARGS(#FUNC, Device_##FUNC)

OCCA_PY_METHODS(
  Device_methods,
  DEVICE_METHOD_NO_ARGS(is_initialized),
  DEVICE_METHOD_NO_ARGS(free),
  DEVICE_METHOD_NO_ARGS(mode),
  DEVICE_METHOD_NO_ARGS(properties),
  DEVICE_METHOD_NO_ARGS(kernel_properties),
  DEVICE_METHOD_NO_ARGS(memory_properties),
  DEVICE_METHOD_NO_ARGS(memory_size),
  DEVICE_METHOD_NO_ARGS(memory_allocated),
  DEVICE_METHOD_NO_ARGS(finish),
  DEVICE_METHOD_NO_ARGS(has_separate_memory_space),
  DEVICE_METHOD_NO_ARGS(create_stream),
  DEVICE_METHOD_NO_ARGS(get_stream),
  DEVICE_METHOD_WITH_KWARGS(set_stream),
  DEVICE_METHOD_NO_ARGS(tag_stream),
  DEVICE_METHOD_WITH_KWARGS(wait_for),
  DEVICE_METHOD_WITH_KWARGS(time_between),
  DEVICE_METHOD_WITH_KWARGS(build_kernel),
  DEVICE_METHOD_WITH_KWARGS(build_kernel_from_string),
  DEVICE_METHOD_WITH_KWARGS(build_kernel_from_binary),
  DEVICE_METHOD_WITH_KWARGS(malloc)
);

static PyTypeObject DeviceType = {
  PyVarObject_HEAD_INIT(NULL, 0)
  "occa.c.Device",                          // tp_name
  sizeof(occa::py::Device),                 // tp_basicsize
  0,                                        // tp_itemsize
  (destructor) Device_dealloc,              // tp_dealloc
  0,                                        // tp_print
  0,                                        // tp_getattr
  0,                                        // tp_setattr
  0,                                        // tp_reserved
  0,                                        // tp_repr
  0,                                        // tp_as_number
  0,                                        // tp_as_sequence
  0,                                        // tp_as_mapping
  0,                                        // tp_hash
  0,                                        // tp_call
  0,                                        // tp_str
  0,                                        // tp_getattro
  0,                                        // tp_setattro
  0,                                        // tp_as_buffer
  Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE, // tp_flags
  "Wrapper for occa::device",               // tp_doc
  0,                                        // tp_traverse
  0,                                        // tp_clear
  0,                                        // tp_richcompare
  0,                                        // tp_weaklistoffset
  0,                                        // tp_iter
  0,                                        // tp_iternext
  Device_methods,                           // tp_methods
  0,                                        // tp_members
  0,                                        // tp_getset
  0,                                        // tp_base
  0,                                        // tp_dict
  0,                                        // tp_descr_get
  0,                                        // tp_descr_set
  0,                                        // tp_dictoffset
  (initproc) Device_init,                   // tp_init
  0,                                        // tp_alloc
  0                                         // tp_new
};

static bool device_has_valid_module() {
  DeviceType.tp_new = PyType_GenericNew;
  return PyType_Ready(&DeviceType) >= 0;
}

static void device_init_module(PyObject *module) {
  Py_INCREF(&DeviceType);
  PyModule_AddObject(module,
                     "Device",
                     (PyObject*) &DeviceType);
}

OCCA_PY_MODULE(device, OCCA_PY_NO_METHODS)
//======================================
